### PR TITLE
Implement s3x objects expiration support

### PR DESCRIPTION
**Pull request by bosher**
#### Submitter checklist

- [ ] Change is code complete and matches issue description.
- [ ] Change is covered by existing or new tests.
- [ ] Change modifies documentation or comments only.

#### Which issue this PR fixes

fixes #

### CODE REVIEW

File: src/ccowhttpd/ccowobj.c [hunk 1]
```diff
@@ -633,8 +634,21 @@ static int on_edgex(ccowobj_handler_t *self, h2o_req_t *req,
 			send_error("Create io", err, req);
 			goto _exit;
 		}
+		// Bucket options
+		BUCKET_OPTIONS *bucket_options = NULL;
+		err = get_bucket_options(tc, self->cid, self->tid, bid, &bucket_options);
+		if (err) {
+			send_error("Get bucket options error", err, req);
```
Comment (dyusupov) [inline src/ccowhttpd/ccowobj.c]: goto _exit?

File: src/ccowhttpd/session.c [hunk 2]
```diff
@@ -310,6 +310,8 @@ session_create(objio_info_t *ci, param *comp, int method_type,
 	}
 
 	// Get create headers
+	ss->ci->expiration = expiration;
```
Comment (dyusupov) [inline src/ccowhttpd/session.c]: can expiration be uninitialized here?

File: src/ccowhttpd/ccowobj.c (hunk 1 above)
Comment (bosher) [reply src/ccowhttpd/ccowobj.c]: Options not available for some reason, so we use default options.

File: src/ccowhttpd/session.c (hunk 2 above)
Comment (bosher) [reply src/ccowhttpd/session.c]: It should be zero even if options are not available.

File: src/ccowhttpd/libauth/bucket_options.c [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+bucket_options_put_ht(BUCKET_OPTIONS *bucket_options)
+{
+	int err = 0;
+	const char *bucket_optionskey = bucket_options_property_string(bucket_options, "bucket_optionskey", NULL);
+	if (bucket_optionskey == NULL) {
+		return EINVAL;
+	}
+
+	log_trace(lg, "ht put bucket_options: %s", bucket_optionskey);
+	bucket_options_delete_ht(bucket_optionskey);
```
Comment (dyusupov) [inline src/ccowhttpd/libauth/bucket_options.c]: should be locked?

File: src/ccowhttpd/libauth/bucket_options.c [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+void
+bucket_options_delete_ht(const char *bucket_optionskey)
+{
+	pthread_mutex_lock(&bucket_options_ht_lock);
+	char buf[MAX_KEY_LEN];
+	size_t ent_size;
+	BUCKET_OPTIONS *bucket_options = hashtable_get(bucket_options_ht, (void *)bucket_optionskey, strlen(bucket_optionskey) + 1, &ent_size);
+
+	if (bucket_options != NULL && ent_size == sizeof(BUCKET_OPTIONS)) {
+		strcpy(buf, bucket_optionskey);
```
Comment (dyusupov) [inline src/ccowhttpd/libauth/bucket_options.c]: can we eliminate strcpy?

File: src/ccowhttpd/libauth/bucket_options.c [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+	struct ccow_metadata_kv *kv = NULL;
+	while ((kv = ccow_lookup_iter(iter, CCOW_MDTYPE_METADATA, pos++))) {
+		if (strcmp(kv->key, RT_SYSKEY_NAME_HASH_ID) == 0) {
+			memcpy(bnhid, kv->value, sizeof(uint512_t));
+			n = 1;
+			break;
+		}
+	}
+
+	ccow_lookup_release(iter);
```
Comment (dyusupov) [inline src/ccowhttpd/libauth/bucket_options.c]: should we cleanup iter on ccow_wait() err?

**Comment by dyusupov** [comment]
:star: PR built and available in a preview environment **nexenta-edgefs-pr-338**

**Comment by dyusupov** [comment]
:star: PR built and available in a preview environment **nexenta-edgefs-pr-338**

**Comment by dyusupov** [comment]
:star: PR built and available in a preview environment **nexenta-edgefs-pr-338**

File: src/ccowhttpd/libauth/bucket_options.c [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+	BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL); //Do not use newlines to flush buffer
+	BIO_read(bio, buffer, strlen(b64message));
+	BIO_free_all(bio);
+
+	return (0); //success
+}
+
+static int
+get_options(ccow_t tc, char *bid, int bid_size, char *oid, int oid_size, char *res) {
```
Comment (dyusupov) [inline src/ccowhttpd/libauth/bucket_options.c]: lets pass len of res and do length check while filling it in

**Comment by dyusupov** [comment]
:star: PR built and available in a preview environment **nexenta-edgefs-pr-338**

**Comment by dyusupov** [comment]
:star: PR built and available in a preview environment **nexenta-edgefs-pr-338**

**Pull request merged**